File: labs/08-function-json-mode/java/Main.java
import com.agentlabs.OpenAiClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.List;
import java.util.Map;

public class Main {

    private static final String EVENT_SCHEMA_JSON = """
            {
              "type": "object",
              "properties": {
                "name": {"type": "string"},
                "date": {"type": "string"},
                "location": {"type": "string"},
                "participants": {"type": "array", "items": {"type": "string"}}
              },
              "required": ["name", "date"]
            }
            """;

    private static final String USER_PROMPT = """
            Extract the event details from this message as JSON:

            "Join us for the AI Engineering Meetup on 2025-09-15 at the Shenzhen Hub.
            Attendees: Alice, Bob, and Carol."
            """.strip();

    public static void main(String[] args) throws Exception {
        OpenAiClient client = new OpenAiClient();
        ObjectMapper mapper = new ObjectMapper();

        Map<String, Object> jsonMode = runMode("JSON mode", () -> extractWithJsonMode(client, mapper), mapper);
        Map<String, Object> functionMode = runMode("Function calling", () -> extractWithFunctionCall(client, mapper), mapper);

        System.out.println("\n" + "=".repeat(40));
        System.out.println("Summary");
        System.out.println("=".repeat(40));
        System.out.println("JSON mode ok: " + jsonMode.get("ok"));
        System.out.println("Function call ok: " + functionMode.get("ok"));

        if (Boolean.TRUE.equals(jsonMode.get("ok")) && Boolean.TRUE.equals(functionMode.get("ok"))) {
            Map<String, Object> j = (Map<String, Object>) jsonMode.get("result");
            Map<String, Object> f = (Map<String, Object>) functionMode.get("result");
            System.out.println("\nBoth approaches returned valid events. "
                    + "Function calling gives you an explicit schema contract; "
                    + "JSON mode is simpler when you only need a shaped text response.");
            System.out.println("Names match: " + (j.get("name").equals(f.get("name")) && j.get("date").equals(f.get("date"))));
        }
    }

    private static Map<String, Object> runMode(String label, Extractor extractor, ObjectMapper mapper) throws Exception {
        System.out.println("\n" + "=".repeat(40) + "\nMode: " + label + "\n" + "=".repeat(40));
        try {
            Map<String, Object> result = extractor.extract();
            System.out.println("\n[" + label + "] parsed event:");
            System.out.println(mapper.writerWithDefaultPrettyPrinter().writeValueAsString(result));
            return Map.of("ok", true, "result", result);
        } catch (Exception exc) {
            System.out.println("\n[" + label + "] FAILED: " + exc.getMessage());
            return Map.of("ok", false, "error", exc.getMessage());
        }
    }

    private static Map<String, Object> extractWithJsonMode(OpenAiClient client, ObjectMapper mapper) throws Exception {
        ObjectNode payload = mapper.createObjectNode();
        ArrayNode messages = mapper.createArrayNode();

        ObjectNode system = messages.addObject();
        system.put("role", "system");
        system.put("content", "You are a helpful parser. Return ONLY a JSON object with keys: name, date, location, participants.");

        ObjectNode user = messages.addObject();
        user.put("role", "user");
        user.put("content", USER_PROMPT);

        payload.set("messages", messages);
        ObjectNode responseFormat = mapper.createObjectNode();
        responseFormat.put("type", "json_object");
        payload.set("response_format", responseFormat);
        payload.put("temperature", 0.0);
        payload.put("max_tokens", 300);

        JsonNode response = client.chatCompletion(payload);
        String raw = client.extractMessage(response).get("content").asText();
        System.out.println("\n[JSON mode] raw output:");
        System.out.println(raw);
        return validateEvent(mapper.readValue(raw, Map.class));
    }

    private static Map<String, Object> extractWithFunctionCall(OpenAiClient client, ObjectMapper mapper) throws Exception {
        ObjectNode tool = mapper.createObjectNode();
        tool.put("type", "function");
        ObjectNode function = mapper.createObjectNode();
        function.put("name", "extract_event");
        function.put("description", "Extract event details from user text.");
        function.set("parameters", mapper.readTree(EVENT_SCHEMA_JSON));
        tool.set("function", function);

        ObjectNode payload = mapper.createObjectNode();
        ArrayNode messages = mapper.createArrayNode();

        ObjectNode system = messages.addObject();
        system.put("role", "system");
        system.put("content", "Use the extract_event tool.");

        ObjectNode user = messages.addObject();
        user.put("role", "user");
        user.put("content", USER_PROMPT);

        payload.set("messages", messages);
        ArrayNode tools = mapper.createArrayNode();
        tools.add(tool);
        payload.set("tools", tools);
        ObjectNode toolChoice = mapper.createObjectNode();
        toolChoice.put("type", "function");
        ObjectNode func = toolChoice.putObject("function");
        func.put("name", "extract_event");
        payload.set("tool_choice", toolChoice);
        payload.put("temperature", 0.0);
        payload.put("max_tokens", 300);

        JsonNode response = client.chatCompletion(payload);
        JsonNode message = client.extractMessage(response);
        JsonNode toolCalls = message.get("tool_calls");
        if (toolCalls == null || toolCalls.isEmpty()) {
            throw new IllegalStateException("Model did not call the extract_event tool");
        }
        String raw = toolCalls.get(0).get("function").get("arguments").asText();
        System.out.println("\n[Function call] raw arguments:");
        System.out.println(raw);
        return validateEvent(mapper.readValue(raw, Map.class));
    }

    private static Map<String, Object> validateEvent(Map<String, Object> data) {
        if (!(data instanceof Map)) {
            throw new IllegalArgumentException("Parsed JSON is not an object");
        }
        List<String> required = List.of("name", "date");
        List<String> missing = required.stream().filter(k -> !data.containsKey(k)).toList();
        if (!missing.isEmpty()) {
            throw new IllegalArgumentException("Missing required fields: " + missing);
        }
        Object participants = data.get("participants");
        if (participants != null && !(participants instanceof List)) {
            throw new IllegalArgumentException("Field 'participants' should be an array");
        }
        return data;
    }

    @FunctionalInterface
    interface Extractor {
        Map<String, Object> extract() throws Exception;
    }
}

File: labs/08-function-json-mode/go/main.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	openaiclient "github.com/ZYZ-Labs/agent-labs/shared/config"
)

var eventSchema = map[string]interface{}{
	"type": "object",
	"properties": map[string]interface{}{
		"name":         map[string]interface{}{"type": "string"},
		"date":         map[string]interface{}{"type": "string"},
		"location":     map[string]interface{}{"type": "string"},
		"participants": map[string]interface{}{"type": "array", "items": map[string]interface{}{"type": "string"}},
	},
	"required": []string{"name", "date"},
}

var userPrompt = `
Extract the event details from this message as JSON:

"Join us for the AI Engineering Meetup on 2025-09-15 at the Shenzhen Hub.
Attendees: Alice, Bob, and Carol."
`

func validateEvent(data map[string]interface{}) (map[string]interface{}, error) {
	if data == nil {
		return nil, fmt.Errorf("parsed JSON is not an object")
	}
	var missing []string
	for _, f := range []string{"name", "date"} {
		if _, ok := data[f]; !ok {
			missing = append(missing, f)
		}
	}
	if len(missing) > 0 {
		return nil, fmt.Errorf("missing required fields: %v", missing)
	}
	for key, value := range data {
		spec, ok := eventSchema["properties"].(map[string]interface{})[key]
		if !ok {
			continue
		}
		if s, ok := spec.(map[string]interface{}); ok && s["type"] == "array" {
			if _, ok := value.([]interface{}); !ok {
				return nil, fmt.Errorf("field '%s' should be an array", key)
			}
		}
	}
	return data, nil
}

func chatCompletionRaw(client *openaiclient.Client, reqBody map[string]interface{}) (map[string]interface{}, error) {
	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest("POST", client.Config.BaseURL+"/chat/completions", bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+client.Config.APIKey)
	resp, err := client.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	respBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(respBytes))
	}
	var result map[string]interface{}
	if err := json.Unmarshal(respBytes, &result); err != nil {
		return nil, err
	}
	return result, nil
}

func extractWithJSONMode(client *openaiclient.Client) (map[string]interface{}, error) {
	resp, err := chatCompletionRaw(client, map[string]interface{}{
		"model": client.Config.Model,
		"messages": []map[string]interface{}{
			{
				"role": "system",
				"content": "You are a helpful parser. Return ONLY a JSON object with keys: name, date, location, participants.",
			},
			{"role": "user", "content": userPrompt},
		},
		"response_format": map[string]string{"type": "json_object"},
		"temperature":     0.0,
		"max_tokens":      300,
		"stream":          false,
	})
	if err != nil {
		return nil, err
	}
	raw := extractContent(resp)
	fmt.Println("\n[JSON mode] raw output:")
	fmt.Println(raw)
	var data map[string]interface{}
	if err := json.Unmarshal([]byte(raw), &data); err != nil {
		return nil, err
	}
	return validateEvent(data)
}

func extractWithFunctionCall(client *openaiclient.Client) (map[string]interface{}, error) {
	tool := map[string]interface{}{
		"type": "function",
		"function": map[string]interface{}{
			"name":        "extract_event",
			"description": "Extract event details from user text.",
			"parameters":  eventSchema,
		},
	}
	resp, err := chatCompletionRaw(client, map[string]interface{}{
		"model": client.Config.Model,
		"messages": []map[string]interface{}{
			{"role": "system", "content": "Use the extract_event tool."},
			{"role": "user", "content": userPrompt},
		},
		"tools":      []map[string]interface{}{tool},
		"tool_choice": map[string]interface{}{"type": "function", "function": map[string]string{"name": "extract_event"}},
		"temperature": 0.0,
		"max_tokens":  300,
		"stream":      false,
	})
	if err != nil {
		return nil, err
	}
	raw := extractFirstToolCallArguments(resp)
	if raw == "" {
		return nil, fmt.Errorf("model did not call the extract_event tool")
	}
	fmt.Println("\n[Function call] raw arguments:")
	fmt.Println(raw)
	var data map[string]interface{}
	if err := json.Unmarshal([]byte(raw), &data); err != nil {
		return nil, err
	}
	return validateEvent(data)
}

func extractContent(resp map[string]interface{}) string {
	choices, _ := resp["choices"].([]interface{})
	if len(choices) == 0 {
		return ""
	}
	choice, _ := choices[0].(map[string]interface{})
	message, _ := choice["message"].(map[string]interface{})
	content, _ := message["content"].(string)
	return content
}

func extractFirstToolCallArguments(resp map[string]interface{}) string {
	choices, _ := resp["choices"].([]interface{})
	if len(choices) == 0 {
		return ""
	}
	choice, _ := choices[0].(map[string]interface{})
	message, _ := choice["message"].(map[string]interface{})
	toolCalls, _ := message["tool_calls"].([]interface{})
	if len(toolCalls) == 0 {
		return ""
	}
	tc, _ := toolCalls[0].(map[string]interface{})
	fn, _ := tc["function"].(map[string]interface{})
	args, _ := fn["arguments"].(string)
	return args
}

func runMode(label string, extractor func(*openaiclient.Client) (map[string]interface{}, error), client *openaiclient.Client) map[string]interface{} {
	fmt.Printf("\n%s\nMode: %s\n%s\n", strings.Repeat("=", 40), label, strings.Repeat("=", 40))
	result, err := extractor(client)
	if err != nil {
		fmt.Printf("\n[%s] FAILED: %v\n", label, err)
		return map[string]interface{}{"ok": false, "error": err.Error()}
	}
	fmt.Printf("\n[%s] parsed event:\n", label)
	b, _ := json.MarshalIndent(result, "", "  ")
	fmt.Println(string(b))
	return map[string]interface{}{"ok": true, "result": result}
}

func main() {
	client := openaiclient.NewClient()

	jsonMode := runMode("JSON mode", extractWithJSONMode, client)
	functionMode := runMode("Function calling", extractWithFunctionCall, client)

	fmt.Println("\n" + strings.Repeat("=", 40))
	fmt.Println("Summary")
	fmt.Println(strings.Repeat("=", 40))
	fmt.Printf("JSON mode ok: %v\n", jsonMode["ok"])
	fmt.Printf("Function call ok: %v\n", functionMode["ok"])

	if jsonMode["ok"].(bool) && functionMode["ok"].(bool) {
		j := jsonMode["result"].(map[string]interface{})
		f := functionMode["result"].(map[string]interface{})
		fmt.Println("\nBoth approaches returned valid events. Function calling gives you an explicit schema contract; JSON mode is simpler when you only need a shaped text response.")
		fmt.Printf("Names match: %v\n", j["name"] == f["name"] && j["date"] == f["date"])
	}
}
